### DISCUSSION TITLE
Why is the code's structure splitted into OrchardCore.XXXX and OrchardCore.XXXX.Abstractions

### DISCUSSION THREAD

**Discussion by HengzheLi** [OrchardCMS/OrchardCore#6988] [General]
I sow almost the orchardcore components whose code are splitted into two projects: OrchardCore.XXXX and OrchardCore.XXX.Abstractions. 

Some components, like logging and localization, do have multiple impletementation. But a lot of components have only one impletementation for its Abstraction. This doubles the project count. 

What is the concept behind this?

**Comment by hishamco** [comment]
The code is splitted into `OrchardCore.XXXX.Abstractions` and `OrchardCore.XXXX` is something align to ASP.NET Core APIs style

> Some components, like logging and localization, do have multiple implementation. But a lot of components have only one implementation for its Abstraction. This doubles the project count.

For Module / Extension that may have multiple implementation, you will have at least one implementation out of the box, such as Localization, `OrchardCore.Localization.Abstractions` and `OrchardCore.Localization.PortableObjects` but you could have something like `OrchardCore.Localization.Json` .. etc. Please have a look to `Microsoft.Extensions.XXX` to see the style in more details

Sometimes you can find `OrchardCore.XXXX.Core` where this the core / default implementation like what we have in Email & Feed modules

**Reply by hishamco** [reply]
FYI this pattern will allow you to extend the framework and replace the default implementation to meet your requirements instead of merge the abstraction and implementation in one project
